src/phantom_wiki/generate_dataset.py

```diff
@@ -0,0 +1,307 @@
+                file.write(article)
+            with open(os.path.join(article_dir, f"{name}_facts.txt"), "w") as file:
+                file.write("\n".join(facts))
+    elif article_format == "json":
+        save_path = os.path.join(output_dir, "articles.json")
+        logging.info(f"Saving articles to: {save_path}")
+        with open(save_path, "w") as file:
+            json.dump(
+                [
+                    {"title": name, "article": article, "facts": facts}
+                    for name, (article, facts) in articles.items()
+                ],
+                file,
+                indent=4,
+            )
+    else:
+        raise ValueError(f"Article format {article_format} not supported!")
+    timings["articles_save"] = time.time() - start
+
+    #
+    # Step 3. Generate question-answer pairs
+    #
+    blue("Generating question answer pairs")
+    start = time.time()
+    # generate question templates with a given depth
+    templates = generate_templates(depth=question_depth)
+    # sample questions for each template (i.e., type)
+    if question_format == "json_by_type":
+        question_dir = os.path.join(output_dir, "questions")
+        logging.info(f"Saving questions to: {question_dir}")
+        os.makedirs(question_dir, exist_ok=True)
+
+    progbar = tqdm(enumerate(templates), desc="Generating questions", total=len(templates))
+
+    # Populate person name bank for the universe. The list is static across generating questions
+    # so create it once and pass it to the question generation function
+    person_name_bank: list[str] = db.get_person_names()
+
+    # Create caches for person -> (attr name, attr value) and person -> (relation, related person) pairs
+    # When we iterate over multiple questions, we can reuse the same cache to avoid recomputing
+    # e.g. "John" -> [("dob", "1990-01-01"), ("job", "teacher"), ("hobby", "reading"),
+    # ("hobby", "swimming"), ...]
+    # NOTE: Invariant: (attr name, attr value) pairs are unique
+    person_name2attr_name_and_val: dict[str, list[tuple[str, str]]] = {}
+    # e.g. "John" -> [("child", "Alice"), ("child", "Bob"), ("friend", "Charlie"), ...]
+    # NOTE: Invariant: (relation, related person) pairs are unique
+    person_name2relation_and_related: dict[str, list[tuple[str, str]]] = {}
+
+    # To store all the questions and queries for all templates
+    all_questions = []
+    all_queries = []
+
+    for i, (question_template, query_template, answer) in progbar:
+        # Reset the seed at the start of each question type
+        # so that sampled questions are the same for each question type
+        rng = np.random.default_rng(seed)
+
+        # To store the questions and queries for the given template
+        questions = []
+        queries = []
+
+        # for _ in range(args.num_questions_per_type):
+        while (
+            len(questions) < num_questions_per_type
```

Comment:
What happens if it's somehow not possible to generate the required `num_questions_per_type`? Should we add some max counter?

Reply:
`sample_question()` tries <100 times to sample a question given a template type, and it usually samples in 1 try, maximum within 2-3 tries. Theoretically it could be possible that some templates cannot be translated to a question. Practically we don't see this.

We don't have those kinds of templates right now, so imo let's keep this as it is and come back to this when we find such impossible templates.

Reply:
Could we then add a note that this is a potential bug (so that it's easy to find later?)
```suggestion
            len(questions) < num_questions_per_type
            # TODO: handle potential edge cases where templates repeatedly fail to generate,
            # resulting in an infinite loop
```